### PR TITLE
Data saving and validation

### PR DESCRIPTION
**Pull request by ezequielsetzesglobant**
A singleton was created as a database to save reservations
A validation of the reservation fields was also added

Pixel
https://user-images.githubusercontent.com/84338513/124149666-31632580-da67-11eb-8b59-992f73dbc707.mp4

Nexus
https://user-images.githubusercontent.com/84338513/124149679-3627d980-da67-11eb-91b0-a14b75c5aaaf.mp4

### CODE REVIEW

File: app/src/main/java/com/example/parkingsystem/mvp/model/ReservationModel.java [hunk 1]
```diff
@@ -1,29 +1,81 @@
...
+import com.example.parkingsystem.mvp.model.reservation.Reservation;
+import com.example.parkingsystem.mvp.model.reservation.ReservationInfomation;
+import java.text.SimpleDateFormat;
+import java.util.Calendar;
 
 public class ReservationModel implements ReservationContract.ReservationModelContract {
 
-    private String date;
-    private String time;
+    private static final String BUTTON_START_DATE = "Start date";
```
Comment (SartiMau) [inline app/src/main/java/com/example/parkingsystem/mvp/model/ReservationModel.java]: I think that all these constants are using to show the toast at the end. this should be in a resource and be called from the view

File: app/src/main/java/com/example/parkingsystem/mvp/model/reservation/ReservationInfomation.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+    private static ReservationInfomation reservationInfomation = null;
+    private static Map<String, List<Reservation>> reservations;
+
+    private ReservationInfomation() {
+        reservations = new HashMap<>();
+    }
+
+    public static ReservationInfomation getInstance() {
+        if (reservationInfomation == null)
```
Comment (SartiMau) [inline app/src/main/java/com/example/parkingsystem/mvp/model/reservation/ReservationInfomation.java]: remember to use the curly brackets pls

File: app/src/main/java/com/example/parkingsystem/mvp/model/reservation/ReservationInfomation.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        return reservationInfomation;
+    }
+
+    public List<Reservation> getReservations(String place) {
+        return reservations.get(place);
+    }
+
+    public void putReservation(Reservation reservation) {
+        List<Reservation> reservationsList = reservations.get(reservation.getPlace());
+        if (reservationsList != null)
```
Comment (SartiMau) [inline app/src/main/java/com/example/parkingsystem/mvp/model/reservation/ReservationInfomation.java]: same here

File: app/src/main/java/com/example/parkingsystem/mvp/presenter/ReservationPresenter.java [hunk 4]
```diff
@@ -14,25 +14,29 @@ public ReservationPresenter(ReservationContract.ReservationModelContract model,
...
+    public void saveReservationTime(int hour, int minute) {
+        model.saveTime(hour, minute);
     }
 
     @Override
-    public void backToMainActivity() {
-        view.finishActivity();
+    public void saveReservationInformation(String securityCode, String place) {
+        String information = model.saveReservation(securityCode, place);
+        if (information != null)
```
Comment (SartiMau) [inline app/src/main/java/com/example/parkingsystem/mvp/presenter/ReservationPresenter.java]: use the curly brackets pls

File: app/src/main/java/com/example/parkingsystem/mvp/model/ReservationModel.java [hunk 5]
```diff
@@ -1,29 +1,81 @@
...
+    private Calendar startDateAndTime = null;
+    private Calendar finishDateAndTime = null;
+    private boolean startTime = false;
+    private boolean finishTime = false;
 
     @Override
-    public void saveDate(String date) {
-        this.date = date;
+    public void saveDate(int year, int month, int dayOfMonth) {
+        if (buttonText.equals(BUTTON_START_DATE)) {
```
Comment (lucianoeli) [inline app/src/main/java/com/example/parkingsystem/mvp/model/ReservationModel.java]: you shouldnt use this metod to distinguish start from end button.

if you think in the future to change the text of the btn, you will have a dependence in your logic.

i think you can use a flag in the model, or a != empty logic for this case

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/parkingsystem/mvp/presenter/ReservationPresenter.java [hunk 6]
```diff
@@ -14,25 +14,29 @@ public ReservationPresenter(ReservationContract.ReservationModelContract model,
...
+            view.finishActivity(information);
+        else
+            view.showError();
```
Comment (lucianoeli) [inline app/src/main/java/com/example/parkingsystem/mvp/presenter/ReservationPresenter.java]: dont like how `model.saveReservatriion(...)` does two things (save on db, and return) think in single responsability.

you should only save, and make a method `model.getReservation(...)` for the verification. This way you are also checking that de reservation does actually been save.

**Review by lucianoeli — CHANGES_REQUESTED** [review]
good job, approve after coments

File: app/src/main/java/com/example/parkingsystem/mvp/model/ReservationModel.java (hunk 5 above)
Comment (ezequielsetzesglobant) [reply app/src/main/java/com/example/parkingsystem/mvp/model/ReservationModel.java]: done

File: app/src/main/java/com/example/parkingsystem/mvp/model/ReservationModel.java [hunk 7]
```diff
@@ -1,29 +1,81 @@
...
+            return START_DATE + startDate + LINE_JUMP +
+                    FINISH_DATE + finishDate + LINE_JUMP +
+                    SECURITY_CODE + reservation.getSecurityCode() + LINE_JUMP +
+                    PLACE + reservation.getPlace();
+        }
+        return null;
+    }
+
+    private boolean validFields(String securityCode, String place) {
+        if (startDateAndTime != null && finishDateAndTime != null && startTime && finishTime && !securityCode.isEmpty() && !place.isEmpty())
```
Comment (lucianoeli) [inline app/src/main/java/com/example/parkingsystem/mvp/model/ReservationModel.java]: small improve could be to use an OR and check if only one is missing to return false.
the order its imoportant too, most common checks to be false, should be at the left. this decreese number of checks

File: app/src/main/java/com/example/parkingsystem/mvp/model/reservation/ReservationInfomation.java (hunk 2 above)
Comment (ezequielsetzesglobant) [reply app/src/main/java/com/example/parkingsystem/mvp/model/reservation/ReservationInfomation.java]: done

File: app/src/main/java/com/example/parkingsystem/mvp/model/reservation/ReservationInfomation.java (hunk 3 above)
Comment (ezequielsetzesglobant) [reply app/src/main/java/com/example/parkingsystem/mvp/model/reservation/ReservationInfomation.java]: done

File: app/src/main/java/com/example/parkingsystem/mvp/presenter/ReservationPresenter.java (hunk 4 above)
Comment (ezequielsetzesglobant) [reply app/src/main/java/com/example/parkingsystem/mvp/presenter/ReservationPresenter.java]: done

File: app/src/main/java/com/example/parkingsystem/mvp/model/ReservationModel.java (hunk 7 above)
Comment (ezequielsetzesglobant) [reply app/src/main/java/com/example/parkingsystem/mvp/model/ReservationModel.java]: done

File: app/src/main/java/com/example/parkingsystem/mvp/presenter/ReservationPresenter.java (hunk 6 above)
Comment (ezequielsetzesglobant) [reply app/src/main/java/com/example/parkingsystem/mvp/presenter/ReservationPresenter.java]: done

File: app/src/main/java/com/example/parkingsystem/mvp/model/ReservationModel.java (hunk 1 above)
Comment (ezequielsetzesglobant) [reply app/src/main/java/com/example/parkingsystem/mvp/model/ReservationModel.java]: done

**Review by lucianoeli — APPROVED** [review]

**Review by SartiMau — APPROVED** [review]

**Review by BQuinteros — APPROVED** [review]

**Review by LuchoElissondoGlobant — APPROVED** [review]

**Review by BQuinteros — APPROVED** [review]

**Review by SartiMau — APPROVED** [review]

**Pull request merged**